README
BT Library README File

26th November, 2012

CONTENTS
    INTRODUCTION
    UPDATING
    COMPILING
    INSTALLING
    NOTES
    BUGS
    ERROR REPORTING
    RELEASE HISTORY
    ACKNOWLEDGEMENTS

INTRODUCTION
    The BT library is a set of C language functions that implement a
    generalised index file capability, based on the B Tree indexing
    scheme. The B Tree was originally described by Bayer and
    McCraight. A B tree is a multi-way balanced tree: i.e. there is
    more than one key per node, and all leaf nodes are the same
    distance from the root node.

    The BT functions implement a 'classical' B tree, not one of the
    later variants (B* or B+ tree).

    The B Tree is stored in a disk file. The file can contain many B
    Trees, each of which is referred to by a unique name assigned by
    the user (or application). The system allows many such files to
    exist.

    In order for BT to function efficiently on many hardware
    platforms, the important variables relating to disk block size,
    maximum number of keys per block, etc. are defined as constants,
    and may be modified prior to compilation (see inc/bc.h).

    BT supports large files (> 2GB), if built with the Large File
    Support (LFS) option.

    This version (5.0.0) is principally a re-write of duplicate key
    support, which removes the restrictions on index file navigation
    in shared mode.

    The BT library has been tested on GNU/Linux (Debian Squeeze), FreeBSD
    9.1, Mac OS X 10.7.5 and Cygwin.

UPDATING
    The BT Recovery program (btr) enables the migration of btree
    index files created in earlier versions.

    If migrating from a pre-3.0 release of the BT library, be aware
    that the function btcerr always zero-pads the srname and msg
    buffers to BT size constants ZRNAMESZ and ZMSGSZ
    respectively. Declaring these arrays to be smaller than the BT
    constants will ensure btcerr acts as a very effective stack
    smasher.

COMPILING
    First, grab the tar file from http://www.hydrus.org.uk and untar
    into a convenient directory, e.g.

        tar xzf bt-5.0.0.tar.gz -C ~/dev

    Since the tar file includes this README file, you've probably done
    this already.

    The key sub-directories under the bt source directory are as follows:
 
        src-lib             C source for BT library routines
        src-main            C source for BT test harness and copy
                            utility
        inc                 C header files
        lib                 holds BT library archive
        samples             simple programs showing library API usage
        Testcases           driver and master output files for testing

    The BT library, a test harness (bt), additional testing tools for
    large files (bigt and bigtdel), recovery tool (btr) and a copy
    utility (kcp) are built by GNU make.  The test harness, bt, will
    be built with readline support, if the make file detects support
    on the build platform. To build the BT library and associated
    programs for up to 2GiB file support, cd to the BT source
    directory and issue the following command:

       make clean && make

    To build the LFS version of BT (supports files >2GiB), issue the
    following command:

       make clean && make LFS=1

    The BT library is a standard UNIX object archive, built by ar.

    It is possible to smoke test the newly built library by:
       
       make test_run

    All tests should pass.  Note that the last test (bigdata) will
    consume some 2.4GB of disk space and take some time, if the LFS
    version of BT is built.

    If you need to rebuild, issue the command "make clean && make".

INSTALLING
    Copy the BT library archive (./lib/libbt.a) to a convenient system
    library location (e.g. /usr/local/lib).  The standard BT header
     files (./inc/{bc.h,bt.h,btree.h}) must also be placed in a local
    system include directory (e.g. /usr/local/include).

    Programs can then be compiled and linked against the library by a
    command of the following form (NB: the define for
    _FILE_OFFSET_BITS may be required when using a BT library built
    with Large File Support):

         cc -o yak -lbt yak.c [-D_FILE_OFFSET_BITS=64]

    Alternatively, just put the include and library files somewhere
    convenient in your local development hierarchy and use matching
    compiler arguments.

NOTES
    Documentation on the BT API is available in docbook format and
    HTML, also downloadable from http://www.hydrus.org.uk.

    BT is dependent on very little, but it does require the gcc
    compiler and GNU make.

BUGS 
    Shared file handling does not operate properly on Cygwin.  The
    timeout alarm never fires, so a process waiting for shared access
    to a index held exclusive by another process will wait until the
    exclusive lock is released.

    On all platforms, shared file handling is clunky.

ERROR REPORTING
    Should you encounter errors during use of the BT library, please
    email mark@hydrus.org.uk with relevant details.

LICENCE
    Copyright (C) 2003, 2004, 2005, 2008, 2010, 2011. 2012  Mark Willson

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
    02111-1307, USA.

RELEASE HISTORY
    15th December,  2003        1.0     First public release
    31st May,       2004        2.0     Major API revision
     8th September, 2004        2.0.1   Bug fix release
     5th October,   2004        2.0.2   Bug fix release
    28th December,  2005        2.0.3   Bug fix release
    10th May,       2008        2.0.4   Minor bug fixes/OpenSolaris tested
     4th June,      2010        3.0.0   Large file support (> 2GiB)
     2nd July,      2010        3.0.1   Added readline support; bug fixes
    10th December,  2010        3.1.0   Support duplicate keys; previous key
    21st December,  2010        3.1.1   Fix broken test scripts
     3rd January,   2011        3.1.2   Shared mode/duplicate key fixes
    24th June       2011        4.0.0   Added recovery capabilities
    26th November   2012        5.0.0   Duplicate key handling re-written


ChangeLog
2014-10-10  Mark Willson  <mark@amber.hydrus.org.uk>

	* bprvky.c: Correct function name in bterr call.

2012-11-22    <GBMAWIL@GB-L-104529>

	* btr.c: Handle invalid source file version in LFS mode. 

2012-11-21    <mark@GB-L-104529>

	* btr.c: Fix seg fault with too few arguments.  Add usage string.

012-11-13  Mark Willson  <mark@aqua>

	* btdata.c: ZDUP recognised as a legitimate data block. Minimum
	segment size stored in infblk[5] in a ZDUP/ZDATA block. Used to
	ensure duplicate key entries are never split, making btr's
	recovery job easier.

2012-11-12  Mark Willson  <mark@aqua>

	* Makefile: Added -lcurses to library list (required for OpenBSD).
	Added btr.h as dependency for btr.c.

	* btr.c: Added processing for duplicate key blocks.  Additional
	statistics displayed.

	* btdupkeys.c: New module for revised duplicate key handling -
	restrictions on duplicate key support and shared mode index
	navigation removed. Index file version bumped to x05; index file
	structure has changed and new block type (ZDUP) introduced.  Many
	changes to existing bt functions to support the new duplicate key
	implementation.

2012-04-11    <mark@GB-L-108394>

	* bterr.c: Make error message for QNOTDA error code less cryptic.

2012-04-10    <mark@GB-L-108394>

	* btdata.c: Check value of key is a valid draddr before use.

2011-06-24  Mark Willson  <mark@amber.hydrus.org.uk>

	* btchgr.c: Added check on block type of candidate root.

2011-06-22  Mark Willson  <mark@amber.hydrus.org.uk>

	* samples: Added directory with some BT Library API examples.

2011-06-13  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c: Added wt command to set write through threshold.

	* bt.h, bputky.c, bremky.c, bdata.c, btthresh.c: Added support for
	block write through when write threshold is reached, as an attempt
	to reduce the index file damage on unexpected btree termination.

2011-05-21  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c, btcmd.c: Add instant quit option to make it easy to create
	damaged index files.

2011-05-02  Mark Willson  <mark@amber.hydrus.org.uk>

	* btr.c: Created initial version.

	* btcrt.c, btcrtr.c: Root block number stored in ZNBLKS info word
	in ZROOT blocks.

	* bterr.c, bsuper.c, bstinf.c, bsetbk.c, bgtinf.c, bc.h: Bumped
	version of BT index to x04.  Added code to detect use of 64 bit
	index file with 32 bit library (and vice versa).

	* btdata.c: Added duplicate draddr detection to brecsz for BT
	recovery.

	* bsptbk.c: Store root block number in ZNBLKS info word of a
	ZINUSE block.  Added to aid BT recovery.

2010-12-31  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c: Improve error handling in next and previous commands.

	* bupdky.c: Prohibit current context operations when index file is
	opened shared and duplicate keys exist.

	* bdelky.c: Prohibit current context operations when index file is
	opened shared and duplicate keys exist.
 
	* btdata.c: Enforced read-only access in data functions.  Revised
	support for btseln and btselp when index file is opened shared and
	the current root supports duplicate keys. Prohibit current context
	operations when shared and duplicate keys exist.

2010-12-06  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c: Fixed problem of erratic error message display.

2010-11-23  Mark Willson  <mark@amber.hydrus.org.uk>

	* btcmd.c: Moved interrupt handling from bt.c, so that exit from
	execute files is handled correctly.

	* balbk1.c: Use brepky to insert new parent when balancing blocks.
	Required due to changed function of bputky.

	* bputky.c: Modified to insert key at beginning of duplicate set.
	If only an rlink is provided (i.e llink ZNULL) key will be placed
	at end of block, to support key demotion.

2010-11-07  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c: New commands to support duplicate key functionality.
	
	* btpos.c: Added API function to allow positioning to start/end of
	index.

2010-11-05  Mark Willson  <mark@amber.hydrus.org.uk>

	* btdups.c: Added to allow dup key support to be set/unset on a
	root basis.

	* binsky.c: Revised to support insertion of duplicate keys, when
	the current root allows duplicate keys.

	* bfndky.c: Revised to support searching when duplicate keys
	exist.  If duplicate keys exist for the required key, the first of
	the duplicate set is returned.

2010-11-02  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c: Added commands to allow testing of previous key
	functionality.

	* bnxtky.c: Modfied to work correctly with bprvky.

	* bprvky.c: Added to allow positioning to previous key.
	

2010-07-23  Mark Willson  <mark@amber.hydrus.org.uk>

	* btcmd.c:  Add support for help on specific command.

2010-07-01  Mark Willson  <mark@amber.hydrus.org.uk>

	* bterr.c: Null'd return arguments before population.

	* bsuper.c: Improved error message on I/O error reading superroot.

	* btdelr.c: Forbid deletion of superroot.

2010-06-18  Mark Willson  <mark@amber.hydrus.org.uk>

	* btcmd.c: Moved bt command handling into this module.  Revamped
	command handling completely.

2010-06-07  Mark Willson  <mark@amber.hydrus.org.uk>

	* bt.c, Makefile: Added support for readline when reading
	interactive commands.  Readline support is only compiled in if the
	readline include files are detected.

	* btcrt.c: Fixed bug - super.smode not set to zero.

2010-06-04  Mark Willson  <mark@amber.hydrus.org.uk>

	* kcp.c: Improved error handling; use btcerr array size constants.

	* bt.c: Added interrupt handling.

	* bigt.c, bigtdel.c: Promoted to src-main; added interrupt handling.

2010-06-02  Mark Willson  <mark@amber.hydrus.org.uk>

	* bterr.c: Returned routine name and error message for btcerr
	constrained to constant-defined sizes (ZRNAMESZ and ZMSGSZ).

2010-05-28  Mark Willson  <mark@amber.hydrus.org.uk>

	* bgtfre.c: Write newly acquired block to file immediately to
	ensure sufficient disk space exists.

	* bsptbk.c: Fix off-by-one error when splitting non-root block
	with even number of keys.


inc/bc.h
/*
 * $Id: bc.h,v 1.37 2012/10/23 19:19:44 mark Exp $
 *
 * Copyright (C) 2003, 2004, 2012 Mark Willson.
 *
 * This file is part of the B Tree library.
 *
 * The B Tree library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * The B Tree library  is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the B Tree library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef _btconst_
#define _btconst_

#include <limits.h>

/*  B tree constants */

/* Support for large files (> 2GB); define macro _FILE_OFFSET_BITS=64 */
#if _FILE_OFFSET_BITS == 64
typedef long long BTint;
#define BTINT_MAX 9223372036854775807LL
#define ZXFMT "%llx"
#define ZINTFMT "%lld"
#define Z20DFMT "%20lld"
#else
typedef int BTint;
#define BTINT_MAX INT_MAX
#define ZXFMT "%x"
#define Z20DFMT "%20d"
#define ZINTFMT "%d"
#endif

#define DEBUG 0

#define TRUE 1
#define FALSE 0

#define LFSHDR 0xFFBADDEF

/* Implementation defined constants */

/* bits per byte */
#define ZBYTEW 8

/* number of bytes per BTint */
#if _FILE_OFFSET_BITS == 64 
#define ZBPW 8
#else
#define ZBPW 4
#endif


/* maximum key size (bytes) */
#define ZKYLEN 32

#if _FILE_OFFSET_BITS==64
/* number of in-memory blocks (assume LFS users are memory rich) */
#define ZMXBLK  16
/* block size in bytes for LFS users (must be power of 2) */
#define ZBLKSZ 8192
#else
/* number of in-memory blocks (3 is the minimum) */
#define ZMXBLK  3
/* block size in bytes (must be power of 2) */
#define ZBLKSZ 1024
#endif

/* number of keys per block */
#define ZMXKEY ((ZBLKSZ-ZBPW-ZINFSZ*ZBPW)/(ZKYLEN+(3*ZBPW))) 
/* number of pad words required */
#define ZPAD ((ZBLKSZ-(ZKYLEN+(3*ZBPW))*ZMXKEY-ZBPW-ZINFSZ*ZBPW)/ZBPW)
/* threshold for block joining */
#define ZTHRES 3
/* number of bt indexes that may be open concurrently */
#define ZMXACT 5
/* number of seconds to wait for locked file to unlock */
#define ZSLEEP 5

/* Message buffer sizes */
#define ZRNAMESZ 16
#define ZMSGSZ 132

/* Indexes into block info words - first ZINFSZ words in every block 
 *
 * ZBTYPE   ZSUPER      ZROOT           ZINUSE       ZFREE   ZDATA 
 * ZMISC    #free blks  dups_allowed    Unused       Unused  bytes free
 * ZNXBLK   free list   data blk list   Parent blk # flink   flink 
 * ZNKEYS   # keys      # keys          # keys       Unused  next free byte 
 * ZNBLKS   # blocks    root blk #      root blk #   Unused  blink
 * ZNXDUP   Unused      Act. dup blk    Unused       Unused  min seg size
 * 
 */

/* ZVERS must be incremented when structure of B Tree index file
 * changes */
#define ZVERS 0x5

/* ZBTYPE and ZBTVER share information word 0 */
#define ZBTYPE 0
#define ZBTVER -1
#define ZMISC  1
#define ZNXBLK 2
#define ZNKEYS 3
#define ZNBLKS 4
#define ZNXDUP 5


/* position constants */
#define ZSTART 1
#define ZEND 2

/* max info word index */
#define ZINFSZ 6

/* block type of root */
#define ZROOT 1
/* block type of inuse */
#define ZINUSE 2
/* block type of free */
#define ZFREE 3
/* block type of data */
#define ZDATA 4
/* block type for duplicate keys */
#define ZDUP 5
/* super block location in index file */
#define ZSUPER 0
/* null block pointer */
#define ZNULL -1

/*  Data block defines:
 *      
 *      Each data record segment is prefixed by ZDOVRH bytes of
 *      information (six bytes for a 32 bit int implementation).
 *      These are used as follows:
 *
 *          Bytes 1 and 2: the size of the data segment in bytes
 *          (maximum size of a data segment is therefore 65536 bytes).
 *          
 *          Bytes 3-6: segment address of the next segment of this
 *          data record (0 if the last (or only) segment)).
 */

#define ZDRSZ 2
#define ZDOVRH  (ZDRSZ+ZBPW)

/* minimum number of bytes for a data segment */
#define ZDSGMN 7    

/* Previous/next key constants */
#define ZPREV 1
#define ZNEXT 2

/* Perform some simple constant consistency checks */

/* Enforce 3 in-memory blocks as minimum
    Need 3 when a root block is split 
*/
#if (ZMXBLK < 3) 
#error Must have minimum of three in-memory blocks
#endif

/* Ensure block joining will take place
*/
#if (ZMXKEY-ZTHRES <= 0)
#error Definition of ZMXKEY and ZTHRES inconsistent
#endif

/*
   Error codes
*/

#define QBLKNR   1
#define QCLSIO   2
#define QCRTIO   3
#define QCPBLK   4

#define QWRBLK   5
#define QRDSUP   6
#define QWRSUP   7
#define QOPNIO   8

#define QRDBLK   9
#define QIXOPN  10
#define QSPLIT  11
#define QINFER  12

#define QNOMEM  13
#define QSTKUF  14
#define QSTKOF  15
#define QBLKFL  16

#define QLOCTB  17
#define QSPKEY  18
#define QWRMEM  19
#define QBALSE  20

#define QDELEX  21
#define QDELER  22
#define QDELRP  23
#define QDEMSE  24

#define QDEMSP  25
#define QJNSE   26
#define QNODEF  27
#define QDELCR  28

#define QBADIX  29
#define QNOBTF  30
#define QINERR  31
#define QBADOP  32

#define QNOACT  33
#define QBADAP  34
#define QBUSY   35
#define QNOTOP  36 /* obsolete in v5 */

#define QNOBLK  37
#define QNEGSZ  38
#define QNOTDA  39
#define QBADCTXT  40

#define QDLOOP  41
#define QUNLCK  42
#define QLRUER  43
#define QDAERR  44

#define QDNEG   45
#define QDUP    46
#define QNOKEY  47
#define QNOWRT  48

#define QNOTFR  49
#define QBADVR  50
#define QDAOVR  51
#define QF2BIG  52

#define QBADAL  53
#define QDRANEG 54

#define QBLKSZERR 55
#define QNODUPS 56

#define QPOSERR 57

#define QNOT64BIT 58
#define Q64BIT 59

#define QNOTDUP 60
#define QDUPSZ 61
#define QBADIR 61

#endif /* _btconst_ */
